server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -391,13 +392,23 @@ public Text readOptionalText() throws IOException {
         if (length == -1) {
             return null;
         }
-        return new Text(readBytesReference(length));
+        byte[] bytes = new byte[length];
+        if (length > 0) {
+            readBytes(bytes, 0, length);
+        }
+        var encoded = new XContentString.EncodedBytes(bytes);
+        return new Text(encoded);
     }
 
     public Text readText() throws IOException {
-        // use StringAndBytes so we can cache the string if it's ever converted to it
+        // use Text so we can cache the string if it's ever converted to it
         int length = readInt();
-        return new Text(readBytesReference(length));
+        byte[] bytes = new byte[length];
+        if (length > 0) {
+            readBytes(bytes, 0, length);
+        }
+        var encoded = new XContentString.EncodedBytes(bytes);
+        return new Text(encoded);
```

Comment:
This looks like repeated code that could benefit from being pulled into a helper method, like we did with `readBytesReference`.